### ISSUE TITLE
get the architecture strings right

### ISSUE THREAD

**Issue by umlaeute** [pure-data/deken#161]
currently there is a big mess on what the architecture string (e.g. `(Linux-x86_64-64)`) really means, thus causing confusion and bugs (pure-data/deken#157 and pure-data/pure-data#185).

i *think* (and @chr15m can probably confirm that, as this was his doing), the *current* meaning or the arch string is:

 - `os` (e.g. `Linux`): this really denotes the *binary format* of the executables, such as `ELF`, `Mach-O` or similar; for practical reasons, this is probably the same as the *kernel* of the OS.
- `machine` (e.g. `x86_64`) denotes the CPU-instruction set.
- `bits` (e.g. `64`) represents the pointer-size of the system, (currently only *64* or *32*).

i think the arch-string needs a re-design - mainly because it never worked.
e.g. all uploaded `OSX/64bit` packages have an architecture-field of `(Darwin-x86_64-32)`, which is just bogus. it has lead to pure-data/pure-data#185 being solved in *Pd-0.48-0* by just ignoring the `bits` field.

also there is a lot of *CPU-compatibility* checking going on, e.g. `amd64` is marked as a superset of `i386` (which is correct, if you only take the CPU-instructions into account). this, in combination with the dropping of the `bits` field in *Pd-0.48-0*, leads to `i386` externals being marked as compatible with `x86_64` systems (e.g. a 64bit Linux OS on an x86_64 CPU), which is wrong as well.

## a new approach
i'd argue for a new approach for the architecture string, inspired by Debian:

- `machine` is really the `OS-architecture`
- `bits` is really the `t_float`-size (it seems that this is what people believe it to be anyhow), and should probably be renamed to `floatsize`)

#### x86_64
`x86_64` denotes a 64bit OS on an x86_64 CPU (aka `amd64`). This automatically implies *64bit pointers*, so there is no need to store this information in the `bits` field.

`x86_64` is compatible *only* with `amd64` (and vice versa).

#### `i386`
this is the standard 32bit OS on an x86_32 architecture (usually this is probably a `i686` CPU).
this might be compatible with `i585` and `i686`, but to avoid cruft, i would just drop those archs.

there's little use in supporting different CPU-instructions on the package level. there might be some extra performance gains when compiling for a 32bit system with amd64 instructions, but since Pd will happily load such an external even if the CPU doesn't support these instructions, this is dangerous at most.
if somebody needs those extra CPU cycles, they probably have to investigate in compiling themselves.

### arch detection
on the plugin-side, we *need* knowledge from the Pd-core about the current architecture.
the Pd-GUI (Wish) simply cannot know, so it shouldn't try.
instead, Pd-core should pass the actually used arch to a `::deken::set_architecture` proc.

i have no clue how to detect the t_float-size via the `deken` packaging utility. (luckily, for now it's save to assume `sizeof(t_float)==32`).

### filename format
changing the semantics of the arch-strings will obviously break the current uploads.
i'd suggest a new filename with a separate extension (`.dek`?).
Being there, I'd also suggest dropping multiple compression formats, and just follow JAVA's jar, OpenDocument and everybody elses example and use `zip` (if we cannot make `tar.gz` cross-platform).

So instead of `bandlimited-v1.0.0-(Darwin-i386-32)(Darwin-x86_64-32)(Linux-amd64-64)-externals.zip` we would then have
~~~
bandlimited-v1.0.0-(Darwin-i386-32)(Darwin-x86_64-32)(Linux-amd64-32).dek
~~~

TODO: what about those `-objects.txt` files?

**Comment by chr15m** [comment]
Yes you are correct @umlaeute that I was trying to achieve the situation where a user would see those packages that will run on their system. This becomes very complicated, especially on ARM platforms where CPUs will run binaries from the previous CPU class as well as their own, and also detection is imperfect.

Agree with sticking with `zip` format.

Is any `t_float` size other than 32 going to happen soon? I would be comfortable dropping the 32 entirely and when 64 builds become available adding to the filename format. Your call of course as I am unlikely to be the one who will write this code. :)

**Comment by umlaeute** [comment]
which i haven't given that much thought in the last months, i wonder what you think about a scheme like:

~~~
LIBRARY[VER](ARCH...).dek
~~~

e.g. 
~~~
bandlimited[v1.0.0](Darwin-i386-32)(Darwin-x86_64-32)(Linux-amd64-32).dek
~~~

this should make the parsing much easier.
(the `-v${VERSION}` part in the current implementation always gives headaches with reliably telling it apart from the library name).
afaict, both `()` and `[]` should be valid on any mainstream filesystem (for the former, we already know that),

**Comment by chr15m** [comment]
Seems like a good idea. Should have been like this from the start. Thanks for giving it thought!

**Comment by umlaeute** [comment]
status report (#177)
- [x] the deken-plugin seems to work
- [x] the deken cmdline tool to create and upload files
- [ ] the deken proxy server (needs to recognize `.dek` files as deken archives; also needs to hide those files from older deken-plugins (which can (now) be identified via the `httpuseragent` field), as they cannot extract `.dek` files)
- [ ] fix puredata.info to allow uploads of filenames with square brackets `[]`.

the last one is a real problem.
while i haven't found anything in WebDAV that would forbid the use of `[]` it seems to be quite impossible on `puredata.info` (i haven't checked the source code yet).
what's even more discouraging, is that even with recent *Plone5* I don't seem to be able to create URLs with brackets (i can upload such files but they get rewritten :frowning_face: )

maybe  @chr15m or somebody else has an idea for a more robust delimiter.

the important features are:
- parsing should be easy :-)
- the version string must be optional
- the version string must not be mistaken for a part of the packagename nor the architecture.
(the former is an issue with the current naming scheme; e.g. `my-values--externals.zip` is ambiguous, it's either a library `my-values` or the `values` version of the `my` library)

**Comment by umlaeute** [comment]
it turns out that the fix on `puredata.info` was pretty simple. i still wonder whether using an *unsafe character* is such a good idea though...

**Comment by chr15m** [comment]
What is the definition of *unsafe character*?

**Comment by umlaeute** [comment]
[RFC1738](https://www.ietf.org/rfc/rfc1738.txt)

**Comment by chr15m** [comment]
D:

**Comment by chr15m** [comment]
It doesn't seem too dangerous to me otherwise we could use something LISPy like `(v...)` where the braces and 'v' character are stripped after parsing:

`bandlimited(v1.0.0)(Darwin-i386-32)(Darwin-x86_64-32)(Linux-amd64-32).dek`

Extracted version string = `1.0.0`.

**Comment by umlaeute** [comment]
ah well, i dislike the `(v...)` approach, unless it goes together with a `(a...)` for the archs (although...this has the possibility to add additional fields in the future).

afaict, i have the `[...]` working for the plugin, the cmdline tool and the server-side so i'll probaby go for that.

on second thought we probably should use `[v3.14]` rather than just `[3.14]` so we *can* extend the format easily without having to look for new delimiters.

how about:

~~~BNF
<filename> ::= <libraryname><options><archs>'.dek'
<libraryname> ::= <word>
<options> ::= | <option><options>
<archs> ::= | <arch><archs>
<option> ::= '['<specifier><word>']'
<arch> ::= '('<os>'-'<cpu>'-'<floatsize>')'
<specifier> ::= 'v' | ...
<os> ::= 'Darwin' | 'Linux' | 'Wndows' | ...
<cpu> ::= 'amd64' | 'i386' | 'arm' | 'armv7' | ...
<floatsize> ::= '32' | '64'
<word> ::= <char><chars>
<chars> ::= |<char><chars>
<char> ::= ... ; valid path-segment characters except '[]()'
~~~

**Comment by katjav** [comment]
@umlaeute, you proposed a consistent extension format in this pull request:

(https://github.com/pure-data/pure-data/pull/297/commits/199ce9ec1b0acb4eba9d819af945d475c7a3d6f4)

Why not use the same in the deken string, like: 

`bandlimited[v1.0.0](l_i386-32)(l_amd64-32)(l_arm-32).dek`

From `s_loader.c` :

> The specific name is the letter b, l, d, or m for  BSD, linux,
> darwin, or microsoft, followed by a more specific string, either "fat" for
> a fat binary or an indication of the instruction set.

Where indication of the instruction set can be `i386`, `amd64`, `arm`, `arm64`, `ppc` (in the pull request).

The os+cpu format is compact, consistent and already (partly) in use. Is there a compelling reason to not use it for both purposes (extension and deken string)? The new extension `.dek` is great by the way.

**Comment by umlaeute** [comment]
@katjav i somehow figured that the deken-filename is more exposed to the user, and the single-character OS-prefixes are a bit...terse.

**Comment by katjav** [comment]
A single character for OS is terse indeed. But what about `i386`, does everyone know what that means? And `32` / `64`, which currently mean 'pointer size' and in the future 'float size'? Could the 'Find externals' window print a header section with explanation? (I'm not even sure that every OSX user understands 'Darwin').

**Comment by umlaeute** [comment]
implemented with `deken-0.3`
